index.js

```diff
@@ -0,0 +1,214 @@
+    tx.hincrby('rewards', address, scheduledRewards)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        score,
+        scheduledRewards: String(scheduledRewards)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.scores)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handlePaidScheduledRewards (req, res, redis, signerAddresses) {
+  const body = JSON.parse(await getRawBody(req, { limit: '1mb' }))
+
+  httpAssert(
+    typeof body === 'object' && body !== null,
+    400,
+    'Request body should be an object'
+  )
+  httpAssert(
+    typeof body.rewards === 'object' && body.rewards !== null,
+    400,
+    '.rewards should be an object'
+  )
+  httpAssert(
+    Object.keys(body.rewards).every(ethers.isAddress),
+    400,
+    'All .rewards keys should be 0x addresses'
+  )
+  httpAssert(
+    Object.values(body.rewards).every(score => {
+      try {
+        return BigInt(score) > 0n
+      } catch {
+        return false
+      }
+    }),
+    400,
+    'All .rewards values should be positive numbers encoded as string'
+  )
+
+  validateSignature(body.signature, body.rewards, signerAddresses)
+
+  const timestamp = new Date()
+  const tx = redis.multi()
+  for (const [address, amount] of Object.entries(body.rewards)) {
+    tx.hincrby('rewards', address, BigInt(amount) * -1n)
+    tx.rpush(
+      'log',
+      JSON.stringify({
+        timestamp,
+        address,
+        scheduledRewards: String(BigInt(amount) * -1n)
+      })
+    )
+  }
+  const updated = await tx.exec()
+
+  json(
+    res,
+    Object.fromEntries(
+      Object
+        .keys(body.rewards)
+        .map((address, i) => [address, String(updated[i * 2][1])])
+    )
+  )
+}
+
+async function handleGetAllScheduledRewards (res, redis) {
+  json(res, await redis.hgetall('rewards'))
+}
+
+async function handleGetSingleScheduledRewards (req, res, redis) {
+  json(res, await redis.hget('rewards', req.url.split('/').pop()))
+}
+
+async function handleGetLog (res, redis) {
+  json(res, (await redis.lrange('log', 0, -1)).map(JSON.parse))
+}
+
+const errorHandler = (res, err, logger) => {
+  if (err instanceof SyntaxError) {
+    res.statusCode = 400
+    res.end('Invalid JSON Body')
+  } else if (err.statusCode) {
+    res.statusCode = err.statusCode
+    res.end(err.message)
+  } else {
+    logger.error(err)
+    res.statusCode = 500
+    res.end('Internal Server Error')
+  }
+
+  if (res.statusCode >= 500) {
+    Sentry.captureException(err)
+  }
+}
+
+export const createHandler = async ({ logger, redis, signerAddresses }) => {
+  assert(logger, '.logger required')
+  assert(redis, '.redis required')
+  assert(signerAddresses, '.signerAddresses required')
+
+  return (req, res) => {
+    const start = new Date()
+    logger.request(`${req.method} ${req.url} ...`)
```

Comment:
Do you have any concerns about producing too many log lines when we log all incoming requests? We had to make this configurable for spark-api. OTOH, spark-stats seems to log all requests, so maybe it's not a problem in spark-rewards either.

Reply:
This is configurable here as well, we can disable it through the `REQUEST_LOGGING` environment variable